### DISCUSSION TITLE
Language mode mistakenly automatically moves C -> C++

### DISCUSSION THREAD

**Discussion by ltWolfik** [microsoft/vscode-cpptools#8848] [Q&A]
In some C projects, for example on [iwd](https://git.kernel.org/pub/scm/network/wireless/iwd.git/), sometimes on opening .h file language mode mistakenly automatically changes to C++.
I disabled `workbench.editor.languageDetection`, but this behavior continues.

Does anyone know how to fix the problem?

VS Code version: 1.63.2
OS: Linux x64 5.16.0-1-default
C/C++ Extension version: v1.7.1

**Comment by sean-mcmanus** [comment]
You could try setting C_Cpp.autoAddFileAssociations to false. If your files end with `.C` we could be associating them with C++ automatically, since normally the capital C is used for C++ and not C.

**Reply by ltWolfik** [reply]
Setting `C_Cpp.autoAddFileAssociations` to `false` does not help.
If i just open `dbus.h` - it opens like C file. Ok. Next i close `dbus.h` and do `Go to Definition` on `#include "src/dbus.h"` - it opens and moves to C++ mode right after.

**Reply by Colengms** [reply]
Hi @ltWolfik .  When a header file is opened, a source file that includes that header is chosen to use for IntelliSense.  If a file that is already open includes the header, that file is used.  If none can be located, the header file is processed by itself.  If your header is included by both C and C++ sources, that might explain the behavior you are seeing.

If you run the `C/C++: Log Diagnostics` command, the information it outputs includes the associations between open header files and source files.  Could you use that to check which source files are being associated with the header in those 2 scenarios?  There may be an issue with one of the sources that are selected being treated as C++ when you are expecting it to be treated as C.  If so, the issue may actually be specific to that source file, not the header.

**Reply by ltWolfik** [reply]
Hi @Colengms.

[Diagnostics-just-open.log](https://github.com/microsoft/vscode-cpptools/files/8068378/Diagnostics-just-open.log)
[Diagnostics-open-by-go-to-definition.log](https://github.com/microsoft/vscode-cpptools/files/8068382/Diagnostics-open-by-go-to-definition.log)

**Reply by ltWolfik** [reply]
I just fixed error in `includePath` configuration and the problem disappeared.
